### PR TITLE
Fix ScanOp test by implementing idx ShflKind in commonShflSync

### PR DESCRIPTION
**Pull request by oplavsic**

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp [hunk 1]
```diff
@@ -348,18 +349,18 @@ static Value commonShflSync(Location loc, ConversionPatternRewriter &rewriter,
...
-    auto shfl = builder.create("ds_bpermute_b32");
-    auto dOpr = builder.newOperand("=v");
-    auto addrOpr = builder.newOperand(permuteAddr, "v");
-    auto aOpr = builder.newOperand(val, "v");
-    (*shfl)(dOpr, addrOpr, aOpr);
+    permute(index, "ds_bpermute_b32");
+    break;
+  }
+  case NVVM::ShflKind::idx:
+    permute(i, "ds_bpermute_b32");
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/Utility.cpp:360]: so shfl.idx means each workitem getting the value from an arbitrary workitem in the same wavefront?

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp [hunk 2]
```diff
@@ -316,6 +308,21 @@ static Value commonShflSync(Location loc, ConversionPatternRewriter &rewriter,
 
 #ifdef USE_ROCM
   GCNBuilder builder;
+
+  auto permute = [&](Value lineId, StringRef permuteInstStr) {
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/Utility.cpp:312]: I'm confused why this is called `lineId` instead of `laneId`?

**Review by zhanglx13 — APPROVED** [review]
LGTM.
And good refactor of the shuffle code.

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp [hunk 3]
```diff
@@ -316,6 +308,21 @@ static Value commonShflSync(Location loc, ConversionPatternRewriter &rewriter,
...
+           permuteInstStr == "ds_bpermute_b32");
+    // Multiple lineId by 4. (More on permute instruction semantics:
+    // https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/instinct-mi200-cdna2-instruction-set-architecture.pdf#page=180
+    Value byteOffset = i32_val(2);
+    Value permuteAddr = shl(lineId, byteOffset);
+    auto shfl = builder.create(permuteInstStr.str());
+    auto dOpr = builder.newOperand("=v");
+    auto addrOpr = builder.newOperand(permuteAddr, "v");
+    auto aOpr = builder.newOperand(val, "v");
+    (*shfl)(dOpr, addrOpr, aOpr);
```
Comment (binarman) [inline lib/Conversion/TritonGPUToLLVM/Utility.cpp:323]: Note: `rocdl` dialect recently got `bpermute` operation.
We can replace this inline asm someday.

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp (hunk 2 above)
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:312]: I think it is coming from [old code](https://github.com/ROCmSoftwarePlatform/triton/pull/391/files#diff-f641b1929956303641de2bed1e9e4019bfe5d4c4c805131f0a2b4ef3110892c9R337)

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp (hunk 3 above)
Comment (jayfurmanek) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:323]: oh yea? I was looking for various shuffles in rocdl, but I didn't see them - I guess I should have been looking for permute instead. That's good.

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp (hunk 2 above)
Comment (jayfurmanek) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:312]: yeah - it was lineId before

**Review by jayfurmanek — APPROVED** [review]

**Review by binarman — APPROVED** [review]
LGTM

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp (hunk 3 above)
Comment (zhanglx13) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:323]: @binarman Do you happen to have the commit in MLIR?

Comment (jayfurmanek) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:323]: https://github.com/llvm/llvm-project/commit/1896096002b75b50d46ee0043c20e90c7e27604a

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp (hunk 1 above)
Comment (oplavsic) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:360]: > so shfl.idx means each workitem getting the value from an arbitrary workitem in the same wavefront?
yes, that is right.

**Pull request merged**